extensions/amp-bind/0.1/bind-expr-impl.jison

```diff
@@ -155,147 +102,198 @@ expr:
+      %}
   | expr '?' expr ':' expr
-      {$$ = $1 ? $3 : $5;}
+      %{
+        $$ = {type: ASTNodeType.TERNARY, args: [$1, $3, $5]};
+      %}
   ;
 
 invocation:
     expr '.' NAME args
       %{
-        $$ = null;
-
-        const obj = toString.call($1);
-
-        const whitelist = functionWhitelist[obj];
-        if (whitelist) {
-          const fn = $1[$3];
-          if (fn && fn === whitelist[$3]) {
-            if (typeCheckArgs($4)) {
-              $$ = fn.apply($1, $4);
-            } else {
-              throw new Error('Unexpected argument type in ' + $3 + '()');
-            }
-            return;
-          }
-        }
-
-        throw new Error($3 + '() is not a supported function.');
+        $$ = {type: ASTNodeType.INVOCATION, args: [$1, $3, $4]};
       %}
   ;
 
 args:
     '(' ')'
-      {$$ = [];}
+      %{
+        $$ = {type: ASTNodeType.ARGS, args: []};
+      %}
   | '(' array ')'
-      {$$ = $2;}
+      %{
+        $$ = {type: ASTNodeType.ARGS, args: [$2]};
+      %}
   ;
 
 member_access:
     expr member
       %{
-        $$ = null;
-
-        if ($1 === null || $2 === null) {
-          return;
-        }
-
-        const type = typeof $2;
-        const isCorrectType = type === 'string' || type === 'number';
-        if (isCorrectType && hasOwnProperty.call($1, $2)) {
-          $$ = $1[$2];
-        }
+        $$ = {type: ASTNodeType.MEMBER_ACCESS, args: [$1, $2]};
       %}
   ;
 
 member:
     '.' NAME
-      {$$ = $2;}
+      %{
+        $$ = {type: ASTNodeType.MEMBER, value: $2};
+      %}
   | '[' expr ']'
-      {$$ = $2;}
+      %{
+        $$ = {type: ASTNodeType.MEMBER, args: [$2]};
+      %}
   ;
 
 variable:
     NAME
-      {$$ = hasOwnProperty.call(yy, $1) ? yy[$1] : null;}
+      %{
+        $$ = {type: ASTNodeType.VARIABLE, value: $1};
+      %}
   ;
 
 literal:
     STRING
-      {$$ = yytext.substr(1, yyleng - 2);}
+      %{
+        $$ = {type: ASTNodeType.LITERAL, value: yytext.substr(1, yyleng - 2)};
+      %}
   | NUMBER
-      {$$ = Number(yytext);}
+      %{
+        $$ = {type: ASTNodeType.LITERAL, value: Number(yytext)};
+      %}
   | TRUE
-      {$$ = true;}
+      %{
+        $$ = {type: ASTNodeType.LITERAL, value: true};
+      %}
   | FALSE
-      {$$ = false;}
+      %{
+        $$ = {type: ASTNodeType.LITERAL, value: false};
+      %}
   | NULL
-      {$$ = null;}
+      %{
+        $$ = {type: ASTNodeType.LITERAL, value: null};
+      %}
   | object_literal
-      {$$ = $1;}
+      %{
+        $$ = $1;
+      %}
   | array_literal
-      {$$ = $1;}
+      %{
+        $$ = $1;
+      %}
   ;
 
 array_literal:
     '[' ']'
-      {$$ = [];}
+      %{
+        $$ = {type: ASTNodeType.ARRAY_LITERAL, args: []};
+      %}
   | '[' array ']'
-      {$$ = $2;}
+      %{
+        $$ = {type: ASTNodeType.ARRAY_LITERAL, args: [$2]};
+      %}
   ;
 
 array:
     expr
-      {$$ = [$1];}
+      %{
+        $$ = {type: ASTNodeType.ARRAY, args: [$1]};
```

Comment:
Looks like object literals are only very marginally faster. Changed to `AstNode` to a class.